### ISSUE TITLE
Mimic LeaderWorkerSet status as StatefulSet or Deployment

### ISSUE THREAD

**Issue by liurupeng** [kubernetes-sigs/lws#96]
For rolling update related status' fields, should we mimic Statefulset or Deployment. My prefer using the StatefulSet one, since it could provide more information.

**Comment by liurupeng** [comment]
@ahg-g @kerthcet

**Comment by ahg-g** [comment]
> For rolling update related status' fields, should we mimic Statefulset or Deployment

+1, we should have this.

**Comment by liurupeng** [comment]
but do you think we should have the same one as StatefulSet or Deployment though @ahg-g

**Comment by ahg-g** [comment]
can you describe what each one covers?

**Comment by kerthcet** [comment]
We're now quite similar to deployment. Glad to have one `observedGeneration` to reflect the current generation.

**Comment by liurupeng** [comment]
In StatefulSet, [Link](https://github.com/kubernetes/kubernetes/blob/b722d017a34b300a2284b890448e5a605f21d01e/pkg/apis/apps/types.go#L247), there are ObservedGeneration, CurrentReplicas, UpdatedReplicas, CurrentRevision, UpdatedRevision etc. 
In Deployment, [Link](https://github.com/kubernetes/kubernetes/blob/b722d017a34b300a2284b890448e5a605f21d01e/pkg/apis/apps/types.go#L507), there are ObservedGeneration, UpdatedReplicas. 

Both could be supported which we could change the CurrentRevision to CurrentTemplateHash or remove that. I mainly want to support the CurrentReplicas and UpdatedReplicas, for now, the UpdatedReplicas==Lws.Spec.Replicas when upgrade finished which make it hard to calculate the upgrade latency (the same in deployment). In statefulSet, when upgrade finished, UpdatedReplcias == 0 and currentReplicas==lws.Spec.Replicas

**Comment by liurupeng** [comment]
@ahg-g and @kerthcet thoughts?

**Comment by ahg-g** [comment]
Adding currentReplicas and changing the semantics of updatedReplicas sounds good to me.

**Comment by ahg-g** [comment]
but we need to do that before the release

**Comment by liurupeng** [comment]
I'll submit a fix today @ahg-g

**Comment by liurupeng** [comment]
actually it's hard to calculate currentReplicas if we don't keep a record of the current template hash somewhere. Does it make sense to only update the semantics of updatedReplicas if we don't add the currentReplicas though

**Comment by kerthcet** [comment]
> actually it's hard to calculate currentReplicas if we don't keep a record of the current template hash somewhere. Does it make sense to only update the semantics of updatedReplicas if we don't add the currentReplicas though

Make sense to me if rollingUpdated finished, set updatedReplicas to zero. Btw, I added the currentReplicas in https://github.com/kubernetes-sigs/lws/pull/98.

**Comment by kerthcet** [comment]
If we set the updatedReplicas to 0 when rolling update finished, then if we have a sts of replicas-10, when two replicas down, what's the value of updatedReplicas right now? eight? then it seems a little weird .. since we jumped from 0 to 8 then to 0 again. 

> calculate the upgrade latency

Can we start when updatedReplicas changes to 0? and end when updatedReplicas = spec.Replicas. I didn't think too deep about this, have no idea whether this works or not.

**Comment by nayihz** [comment]
> If we set the updatedReplicas to 0 when rolling update finished, then if we have a sts of replicas-10, when two replicas down, what's the value of updatedReplicas right now? eight?

Yes. `updatedReplicas` will change when pod restart.

> calculate the upgrade latency

There are `currentRevision` and `updateRevision` to record the template hash. Maybe we can use them to calculate the upgrade latency.

**Comment by kerthcet** [comment]
We decided to add a new condition type or progressing reason for record.

**Comment by ahg-g** [comment]
I just noticed that `UpdatedReplicas` and `Replicas` semantics are not consistent with Statefulset.

In LWS, UpdatedReplicas is the number of replicas that are ready and updated, while in StatefulSet it is the number of updated replicas whether they are ready or not (i.e., created, but not necessarily ready).

In LWS, Replicas is the number of replicas we aim to create (because we set it to leaderSts.Spec.Replicas); while in StatefulSet it is the number of Replicas we created so far (ready or not). I think we can fix this by setting it to `leaderSts.Status.Replicas` instead of `leaderSts.Spec.Replicas`
